src/lib/updating-element.ts

```diff
@@ -0,0 +1,418 @@
+    const info = this._classProperties[name];
+    return info && info.attributeName || name.toLowerCase();
+  }
+
+  /**
+   * Called when a property value is set to determine if the value should change.
+   * This uses the `shouldChange` property option for the given property `name`.
+   */
+  // tslint:disable-next-line no-any
+  protected static propertyShouldChange(name: string, value: any, old: any) {
+    const info = this._classProperties[name];
+    const fn = info && info.shouldChange || identity;
+    return fn(value, old);
+  }
+
+  /**
+   * Called to deserialize an attribute value to a property value.
+   * This uses the `fromAttribute` property option for the given property `name`.
+   */
+  // tslint:disable-next-line no-any
+  protected static propertyValueFromAttribute(name: string, value: string) {
+    const info = this._classProperties[name];
+    const fromAttribute = info && info.fromAttribute;
+    return fromAttribute ? fromAttribute(value) : value;
+  }
+
+  /**
+   * Called to serialize an property value to an attribute value. If this
+   * returns undefined, the property will *not* be reflected to an attribute.
+   * This uses the `toAttribute` property option for the given property `name`.
+   */
+  // tslint:disable-next-line no-any
+  protected static propertyValueToAttribute(name: string, value: any) {
+    const info = this._classProperties[name];
+    const toAttribute = info && info.toAttribute;
+    if (toAttribute !== undefined) {
+      return toAttribute(value);
+    }
+  }
+
+  constructor() {
+    super();
+    this.initialize();
+  }
+
+  /**
+   * Performs element initialization. By default this calls `createRoot` to
+   * create the element `root` node and captures any pre-set values for
+   * registered properties.
+   */
+  initialize() {
+    this.createRoot();
+    // save instance properties.
+    for (const p in (this.constructor as typeof UpdatingElement)._classProperties) {
+      if (this.hasOwnProperty(p)) {
+        // tslint:disable-next-line no-any
+        const me = (this as any);
+        const value = me[p];
+        delete me[p];
+        this._instanceProperties[p] = value;
+      }
+    }
+  }
+
+  /**
+   * Implement to customize where the element's template is rendered by
+   * returning an element into which to render. By default this creates
+   * a shadowRoot for the element. To render into the element's childNodes,
+   * return `this`.
+   * @returns {Element|DocumentFragment} Returns a node into which to render.
+   */
+  protected createRoot() {
+    this.root = this.attachShadow({mode : 'open'});
+  }
+
+  /**
+   * Calls `enableUpdate` and uses ShadyCSS to keep element DOM updated.
+   */
+  connectedCallback() {
+    if (window.ShadyCSS && this._hasUpdated) {
+      window.ShadyCSS.styleElement(this);
+    }
+    this.enableUpdate();
+  }
+
+  /**
+   * Call to enable element updating. This is called in `connectedCallback`
```

Comment:
What does "enable element updating" mean?

Why would it be desired to call sooner?